### ISSUE TITLE
New Rule Idea: Don't use synchronous I/O operations

### ISSUE THREAD

**Issue by zsolt-kolbay-sonarsource** [SonarSource/sonar-dotnet#8219]
**TITLE**
Don't use synchronous I/O operations

**WHY IS THIS AN ISSUE**
Using synchronous I/O operations in applications, especially in environments where scalability and responsiveness are critical, can lead to performance bottlenecks. When a synchronous I/O operation is invoked, it blocks the executing thread until the operation completes. During this time, the thread is not available to perform any other tasks. This can cause the application to become unresponsive, especially under heavy loads or when the I/O operation takes a significant amount of time.

In contrast, asynchronous methods that use [overlapped I/O operations](https://en.wikipedia.org/wiki/Overlapped_I/O) allow the executing thread to continue processing other tasks while the I/O operation is in progress. This ensures better utilization of resources, responsiveness, and scalability of the application.

**NONCOMPLIANT CODE EXAMPLE**

```cs
string content = "This is some content";
File.WriteAllText("path_to_file.txt", content);  // Noncompliant
```

```cs
string content = "This is some content";
await Task.Run(() => File.WriteAllText("path_to_file.txt", content));  // Noncompliant - it will be executed on a thread from the thread pool, but that thread will still be blocked by the I/O operation
```

**COMPLIANT CODE EXAMPLE**

```cs
string content = "This is some content";
await File.WriteAllTextAsync("path_to_file.txt", content);
```

**EXCEPTIONS**

The rule should not raise warnings in command-line applications (Console Applications) or Class Library projects.

**ANALYZER IMPLEMENTATION**

Look for the following method invocations of using file, socket, named pipes, HTTP, etc. operations. 
e.g. for `System.IO.File` the following methods all have async counterparts:
- `File.AppendAllLines`
- `File.AppendAllText`
- `File.ReadAllBytes`
- `File.ReadAllText`
- `File.ReadLines`
- `File.WriteAllBytes`
- `File.WriteAllLines`
- `File.WriteAllText`
It's going to be quite a long list 😎
When a call to any of these methods is encountered, raise a warning and advise to use the `{method_name}Async()` instead.
For each method the analyzer needs to check if the async version is supported by the given runtime before raising an issue (e.g. dont' recommend the use of File.[WriteAllTextAsync](https://learn.microsoft.com/en-us/dotnet/api/system.io.file.writealltextasync) in a .NET Framework 4.8 project, because the method is only available for projects built under .NET Core 2.0 or newer frameworks).

**BENEFITS**
- low hanging fruit
- great educational value (a lot of developers aren't aware how async I/O works)
- concerns other languages/runtimes, not just .NET

**RESOURCES**
- https://en.wikipedia.org/wiki/Overlapped_I/O
- https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/async-scenarios#recognize-cpu-bound-and-io-bound-work
- https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/

**Comment by zsolt-kolbay-sonarsource** [comment]
Superseded by S6966.
